apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -0,0 +1,254 @@
+defmodule Champions.Battle.Simulator do
+  @moduledoc """
+  Runs battles.
+
+  Units have stats that are calculated on battle start (Attack, Max Health, Armor), as well as two skills. The ultimate
+  has no cooldown and it's cast whenever a unit reaches 500 energy. Energy is gained whenever the target attacks.
+  The primary skill has a cooldown and it's cast when it's available if the ultimate is not.
+
+  Skills possess effects that affect other units' or oneself's stats (the one defined in the effect's `stat_affected` field).
+
+  They have different application types (checked are implemented):
+  [x] Instant - Applied once, irreversible.
+  [ ] Permanent - Applied once, is stored in the unit so that it can be reversed (with a dispel, for example)
+  [ ] Duration - Applied once and reverted once its duration ends.
+  [ ] Periodic - Applied many times every x number of steps for a total duration of y steps
+
+  They also have different targeting strategies:
+  [x] Random
+  [ ] Nearest
+  [ ] Furthest
+  [ ] Min Health
+  [ ] Max Health
+  [ ] Min Shield
+  [ ] Max Shield
+  [ ] Frontline - Heroes in slots 1 and 2
+  [ ] Backline - Heroes in slots 2 to 4
+  [ ] Factions
+  [ ] Classes
+
+  And different ways in which their amount is interpreted:
+  [x] Additive
+  [x] Multiplicative
+  [x] Additive & based on stat - The amount is a % of one of the caster's stats
+  [ ] Multiplicative & based on stat?
+
+  Two units can attack the same unit at the same time and over-kill them. This is expected behavior that results
+  from having the battle be simultaneous.
+
+  """
+  alias GameBackend.Units.Skills.Effect
+  alias GameBackend.Units.Skill
+  alias Champions.Units
+  alias GameBackend.Units.Unit
+
+  require Logger
+
+  @maximum_steps 500
+
+  @doc """
+  Runs a battle between two teams.
+  Teams are expected to be lists of units with their character and their skills preloaded.
+
+  Returns `:team_1`, `:team_2`, `:tie` or `:timeout`.
+
+  ## Examples
+
+      iex> team_1 = Enum.map(user1.units, GameBackend.Repo.preload([character: [:basic_skill, :ultimate_skill]]))
+      iex> team_2 = Enum.map(user2.units, GameBackend.Repo.preload([character: [:basic_skill, :ultimate_skill]]))
+      iex> run_battle(team_1, team_2)
+      :team_1
+  """
+  def run_battle(team_1, team_2, seed \\ 1) do
+    :rand.seed(:default, seed)
+    team_1 = Enum.into(team_1, %{}, fn unit -> create_unit_map(unit, 1) end)
+    team_2 = Enum.into(team_2, %{}, fn unit -> create_unit_map(unit, 2) end)
+    units = Map.merge(team_1, team_2)
+
+    # The initial_step_state is what allows the battle to be simultaneous. If we refreshed the accum on every action,
+    # we would be left with a turn-based battle. Instead we take decisions based on the state of the battle at the beggining
+    # of the step regardless of the changes that happened "before" (execution-wise) in this step.
+    Enum.reduce_while(1..@maximum_steps, units, fn step, initial_step_state ->
```

Comment:
Maybe not in this pr but we could replace this `reduce_while` to another function with 3 parameters

`simulate_step(units, winner, steps)`

the base case would be 
`defp simulate_steps(units, winner, step) when not is_nil(winner) or step >= @maximum_steps `

Reply:
Great! We'll create a ticket to refactor this in the short term.

Reply:
https://github.com/lambdaclass/champions_of_mirra/issues/188